### ISSUE TITLE
BUG: iloc fills multiple columns, if columns have duplicate names

### ISSUE THREAD

**Issue by henhuy** [pydata/pandas#12991]
Creating a DataFrame two columns given duplicate names, changing data via indexer changes both columns:

```python
a = pandas.DataFrame(index=['a', 'b', 'c'], columns=['d', 'e', 'd']).fillna(0)
a.iloc[:, 0]['a'] = 3
```
Gives:
    d  e  d
a  3  0  3
b  0  0  0
c  0  0  0

Instead, it should only edit first column:
    d  e  d
a  3  0  0
b  0  0  0
c  0  0  0

INSTALLED VERSIONS
------------------
commit: None
python: 3.4.3.final.0
python-bits: 64
OS: Linux
OS-release: 3.13.0-68-generic
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: de_DE.UTF-8

pandas: 0.17.1
nose: 1.3.6
pip: 1.5.4
setuptools: 3.3
Cython: None
numpy: 1.10.1
scipy: 0.13.3
statsmodels: None
IPython: 3.1.0
sphinx: 1.3.1
patsy: None
dateutil: 2.4.2
pytz: 2015.7
blosc: None
bottleneck: None
tables: 3.1.1
numexpr: 2.2.2
matplotlib: 1.3.1
openpyxl: None
xlrd: 0.9.3
xlwt: 0.8.0
xlsxwriter: 0.7.7
lxml: 3.3.3
bs4: 4.2.1
html5lib: 0.999
httplib2: 0.8
apiclient: None
sqlalchemy: 1.0.4
pymysql: None
psycopg2: 2.6.1 (dt dec pq3 ext lo64)
Jinja2: None

**Comment by jreback** [comment]
This is a tricky bug, though you are chained indexing here, so no guarantees.

Instead do this.
```
In [7]: a.iloc[a.index.get_loc('a'), 0] = 3

In [8]: a
Out[8]: 
   d  e  d
a  3  0  0
b  0  0  0
c  0  0  0
```

**Comment by henhuy** [comment]
That's what I did right now - thanks!


-------------------------------------------
Dipl.-Ing. Hendrik Huyskens
Researcher
Off-Grid Systems

Reiner Lemoine Institut gGmbH
Ostendstraße 25
12459 Berlin
Tel.:    +49 30 53042012
Fax.:   +49 30 53042010
Email: hendrik.huyskens@rl-institut.de<mailto:christiane.basikow@rl-institut.de>
www.reiner-lemoine-institut.de<http://www.reiner-lemoine-institut.de/>

Geschäftsführerin Dr. Kathrin Goldammer
Amtsgericht Berlin-Charlottenburg, HRB 124659 B
________________________________
Von: Jeff Reback <notifications@github.com>
Gesendet: Dienstag, 26. April 2016 15:02
An: pydata/pandas
Cc: Hendrik Huyskens; Author
Betreff: Re: [pydata/pandas] BUG: iloc fills multiple columns, if columns have duplicate names (#12991)


This is a tricky bug, though you are chained indexing here, so no guarantees.

Instead do this.

In [7]: a.iloc[a.index.get_loc('a'), 0] = 3

In [8]: a
Out[8]:
   d  e  d
a  3  0  0
b  0  0  0
c  0  0  0


-

**Comment by toobaz** [comment]
A "more licit" example of the bug in action:

``` python
df1 = DataFrame([{'A': None, 'B': 1}, {'A': 2, 'B': 2}])
df2 = DataFrame([{'A': 3, 'B': 3}, {'A': 4, 'B': 4}])
df = concat([df1, df2], axis=1)
df.iloc[0, 0] = 15
df
```

**Comment by toobaz** [comment]
> A "more licit" example of the bug in action:

For reference: it is actually not the same bug, since it is fixed by #17163 , while the original example by the submitter is not.

I still think, however, that there is some way to expose the (original) bug without chained assigning.

**Comment by MohakGangwani** [comment]
Wouldn't it be better if pandas didn't allow us to have duplicate column names ? Having columns with same name would rather create confusion for us.

**Comment by toobaz** [comment]
> Having columns with same name would rather create confusion for us.

Indeed, not having duplicate column names is a good idea.

Enforcing this in pandas (and doing it consistently - that is, also on rows) would break a lot of code.

**Comment by mitar** [comment]
> would break a lot of code.

But breaking it loudly and consistently is maybe better than breaking some of it silently.

**Comment by toobaz** [comment]
> But breaking it loudly and consistently is maybe better than breaking some of it silently.

Indeed, the safest way to break buggy code loudly and consistently is, always, to just break all code.

**Comment by mitar** [comment]
Not having duplicate column names would also help with sklearn, which is planing to introduce features operating on pandas and column names, but require column names to be unique to work. So having such limit in pandas dataframes would not make you be surprised when you pass it to sklearn and it complains.

**Comment by toobaz** [comment]
> sklearn, which is planing to introduce features operating on pandas and column names, but require column names to be unique to work

For sklearn devs, checking if an index is unique is as simple as checking ``.is_unique``.

For sklearn users, without context I don't know if what you describe is a legitimate need or just an implementation limit, but in any case, it will just have to be documented in sklearn and shouldn't affect the many non-sklearn users.

In any case, I think this is vastly off topic here, as this is a very specific bug caused by a code path which should be rewritten anyway.

**Comment by jbrockmendel** [comment]
@toobaz's example now works correctly, and the OP example is now a no-op, i suspect bc of copy-on-write changing how the chained-indexing operates.  Closing.
